### PR TITLE
Add `s` field to `SurrogateGenerator` to hold surrogates.

### PR DESCRIPTION
**Pull request by kahaaga**
Addresses https://github.com/JuliaDynamics/TimeseriesSurrogates.jl/issues/95 and implements various performance boosts. There were a lot of variables that could be initialized during construction of the `SurrogateGenerator`s. Also, `shuffle` is inefficient compared to `StatsBase.sample`, so swapped these where possible. Additional speed-ups and less memory allocation is probably still possible for many methods, by switching broadcasting operations to specialized for-loops, but I haven't optimized to that extent everywhere. 

## TODO:

Performance is measured on 5000 points long time series (4D datasets for multidimensional methods) with `BenchmarkTools.@btime`.  

- One exception is for `PseudoPeriodicTwin`, where time series used had 1000 points, because initialization freezes due to too many points. This method needs to be heavily optimized.
- Another exception is for `IrregularLombScargle`, where I used time series consisting of 200 points to test (the method is slow). For this method, the bottleneck is computing the Lomb-Scargle periodogram during the loop, so I haven't bothered micro optimizing anything.

- Some methods were already optimized. For those methods,  just had to pre-allocate `s` in the `SurrogateGenerator` instead of in the `init` field. No significant performance increases are seen for those methods.
- The wavelet-based (`WLS`) surrogates still allocate a lot. I think it should be possible to make them more efficient by writing specialized methods (e.g. `(sg::SurrogateGenerator{<:WLS{IAAFT}})()` in addition to the generic `(sg::SurrogateGenerator{<:WLS})()`) that pre-allocate relevant arrays. However, I don't have time for that at the moment, so I'll save that for a later optimization round.

| Status | Method | Old performance | New performance |
| --- | --- | --- | --- |
|  | API |  |  |
| ✅ | `AutoRegressive` | `32.875 μs (2 allocations: 39.17 KiB)`  | `32.000 μs (0 allocations: 0 bytes)` |
| ✅ | `RandomFourier` | `110.458 μs (12 allocations: 176.22 KiB)` | `57.208 μs (4 allocations: 58.72 KiB)` |
| ✅ | `AAFT` | `457.208 μs (17 allocations: 254.45 KiB)` | `257.917 μs (5 allocations: 58.73 KiB)` |
| ✅ | `IAAFT` | `14.009 ms (99689 allocations: 8.83 MiB)` | `1.022 ms (35 allocations: 298.06 KiB)` |
| ✅ | `TFTS` | `143.375 μs (20 allocations: 196.91 KiB)` | `144.750 μs (20 allocations: 196.91 KiB)` |
| ✅ | `TAAFT` | `346.875 μs (23 allocations: 236.03 KiB)` | `146.416 μs (20 allocations: 196.91 KiB)`  |
| ✅ | `TFTDRandomFourier` | `146.834 μs (22 allocations: 235.61 KiB)` | `140.167 μs (14 allocations: 78.56 KiB)` |
| ✅ | `CycleShuffle` | `18.666 μs (0 allocations: 0 bytes)` | `18.541 μs (0 allocations: 0 bytes)` |
| ✅ | `BlockShuffle` | ` 5.817 μs (22 allocations: 80.66 KiB)` | `1.154 μs (0 allocations: 0 bytes)`  |
| ✅ | `CircShift` | `997.375 ns (2 allocations: 39.11 KiB)` | `442.010 ns (0 allocations: 0 bytes)` |
| ✅ | `IrregularLombScargle` | `2.462 s (4551657 allocations: 2.51 GiB)` | `2.462 s (4511722 allocations: 2.51 GiB)` |
| ✅ | `ShuffleDimensions` | `154.250 μs (5002 allocations: 390.67 KiB)` | `149.708 μs (5000 allocations: 234.38 KiB)` |
| ✅ | `RandomShuffle` | `24.791 μs (6 allocations: 117.33 KiB)` | `27.625 μs (4 allocations: 78.22 KiB)` |
| ✅ | `WLS` (timing with IAAFT) | `1.743 s (3838343 allocations: 7.01 GiB)` | `1.460 s (364905 allocations: 6.74 GiB)` |
| ✅ | `PseudoPeriodicTwin` | `423.125 μs (4 allocations: 15.84 KiB)` | `424.250 μs (3 allocations: 96 bytes)` |
| ✅ | `PseudoPeriodic` | `141.478 ms (14999 allocations: 191.12 MiB)` | `141.452 ms (14999 allocations: 191.12 MiB)` |

### CODE REVIEW

File: docs/make.jl [hunk 1]
```diff
@@ -36,16 +36,16 @@ PAGES = [
...
-        "Wavelet-based" => "constrained/wls.md",
-        "Pseudo-periodic twin" => "constrained/ppts.md",
-        "Wavelet-based" => "constrained/wls.md",
-        "Multidimensional surrogates" => "constrained/multidim.md",
-        "Surrogates for irregular timeseries" => "constrained/irregular_surrogates.md",
-        "Surrogates for nonstationary timeseries" => "constrained/nonstationary.md"
+        "Shuffle-based" => "methods/randomshuffle.md",
+        "Fourier-based" => "methods/fourier_surrogates.md",
+        "Amplitude-adjusted FT" => "methods/amplitude_adjusted.md",
+        "Pseudo-periodic" => "methods/pps.md",
```
Comment (Datseris) [inline docs/make.jl:42]: I think you need to refresh to current master or something like that; these commits really shouldn't be here

Comment (kahaaga) [reply docs/make.jl:42]: Not sure what happened here. Somehow, I didn't get the latest commits to master locally when fetching. Will clean up this.

**Comment by Datseris** [comment]
@kahaaga why is everything in new files? Seen the git diff would make this much easier to review.

**Comment by kahaaga** [comment]
> @kahaaga why is everything in new files? Seen the git diff would make this much easier to review.

@Datseris  I'm just keeping both versions while working on optimizations, and have a separate timing script for each method as I do that. Makes it easier to keep track. The final PR, when removing draft status, will have the same files as before.

**Comment by Datseris** [comment]
Ah okay, that makes sense

**Comment by kahaaga** [comment]
I'm getting quite significant speed-ups for `iaaft` after I did some optimizations.


```julia
using TimeseriesSurrogates, BenchmarkTools, Random

rng = MersenneTwister(1234)
rng2 = MersenneTwister(1234)
x = random_cycles(rng, periods = 90)
iaaft = surrogenerator(x, IAAFT(M = 50), rng)
iaaft2 = surrogenerator(x, IAAFT2(M = 50), rng2)
iaaft(); iaaft2();

# Old
julia> iaaft(); @btime sg() setup = (sg = $iaaft);
  16.863 ms (92621 allocations: 9.57 MiB)

# New
julia> iaaft2(); @btime sg() setup = (sg = $iaaft2);
  3.888 ms (75 allocations: 570.06 KiB)
```

**Comment by Datseris** [comment]
@kahaaga that is great! To not get notifications all the time here for each pushed update, I'm unsubscribing from updates. Please do tag me explicitly using `@` when the PR is ready for review and I will of course then do that!

**Comment by kahaaga** [comment]
> @kahaaga that is great! To not get notifications all the time here for each pushed update, I'm unsubscribing from updates. Please do tag me explicitly using @ when the PR is ready for review and I will of course then do that!

Yes, sorry, I forgot to mark the PR as draft when initially making it. All maintainers get notifications by default when a regular PR is opened. To deactivate notifications for all maintainers, the PR must be marked as draft initially.

**Comment by kahaaga** [comment]
Closed with #107

**Pull request closed without merging**